examples/step-12/step-12.cc

```diff
@@ -253,247 +290,168 @@ namespace Step12
-  // base class for both FEFaceValues and FESubfaceValues, in order to get
-  // access to normal vectors.
-  template <int dim>
-  void AdvectionProblem<dim>::integrate_boundary_term(DoFInfo & dinfo,
-                                                      CellInfo &info)
+  void AdvectionProblem<dim>::assemble_system()
   {
-    const FEValuesBase<dim> &fe_face_values = info.fe_values();
-    FullMatrix<double> &     local_matrix   = dinfo.matrix(0).matrix;
-    Vector<double> &         local_vector   = dinfo.vector(0).block(0);
+    typedef decltype(dof_handler.begin_active()) Iterator;
+    BoundaryValues<dim>                          boundary_function;
+
+    // This is the function that will be executed for each cell.
+    auto cell_worker = [&](const Iterator &  cell,
+                           ScratchData<dim> &scratch_data,
+                           CopyData &        copy_data) {
+      const unsigned int n_dofs = scratch_data.fe_values.get_fe().dofs_per_cell;
+      copy_data.reinit(cell, n_dofs);
+      scratch_data.fe_values.reinit(cell);
+
+      const auto &q_points = scratch_data.fe_values.get_quadrature_points();
+
+      const FEValues<dim> &      fe_v = scratch_data.fe_values;
+      const std::vector<double> &JxW  = fe_v.get_JxW_values();
+
+      // We solve a homogeneous equation, thus no right hand side shows up in
+      // the cell term.  What's left is integrating the matrix entries.
+      for (unsigned int point = 0; point < fe_v.n_quadrature_points; ++point)
+        {
+          auto beta_q = beta(q_points[point]);
+          for (unsigned int i = 0; i < n_dofs; ++i)
+            for (unsigned int j = 0; j < n_dofs; ++j)
+              {
+                copy_data.cell_matrix(i, j) +=
+                  -beta_q                      // -\beta
+                  * fe_v.shape_grad(i, point)  // \nabla \phi_i
+                  * fe_v.shape_value(j, point) // \phi_j
+                  * JxW[point];                // dx
+              }
+        }
+    };
+
+    // This is the function called for boundary faces and consists of a normal
+    // integration using FeFaceValues. New is the logic to decide if the term
+    // goes into the system matrix (outflow) or the right-hand side (inflow).
+    auto boundary_worker = [&](const Iterator &    cell,
+                               const unsigned int &face_no,
+                               ScratchData<dim> &  scratch_data,
+                               CopyData &          copy_data) {
+      scratch_data.fe_interface_values.reinit(cell, face_no);
+      const FEFaceValuesBase<dim> &fe_face =
+        scratch_data.fe_interface_values.get_fe_face_values(0);
+
+      const auto &q_points = fe_face.get_quadrature_points();
+
+      const unsigned int n_facet_dofs = fe_face.get_fe().n_dofs_per_cell();
+      const std::vector<double> &        JxW     = fe_face.get_JxW_values();
+      const std::vector<Tensor<1, dim>> &normals = fe_face.get_normal_vectors();
```

Comment:
No humans would ever have come up with the indentation of these three lines :-(

Reply:
Don't get me started with clang-format again, please. Weak forms are also butchered.